flang/lib/Lower/CallInterface.cpp

```diff
@@ -486,3 +501,64 @@ Fortran::lower::CallInterface<T>::getResultType() const {
 
 template class Fortran::lower::CallInterface<Fortran::lower::CalleeInterface>;
 template class Fortran::lower::CallInterface<Fortran::lower::CallerInterface>;
+
+/// Build signature from characterstics when there is no Fortran entity to
+/// associate with the arguments (i.e, this is not a call site or a procedure
+/// declaration. This is needed when dealing with function pointers/dummy
+/// arguments.
+
+class SignatureBuilder;
+template <>
+struct Fortran::lower::PassedEntityTypes<SignatureBuilder> {
+  using FortranEntity = FakeEntity;
+  using FirValue = int;
+};
+
+class SignatureBuilder
+    : public Fortran::lower::CallInterface<SignatureBuilder> {
+public:
+  SignatureBuilder(const Fortran::evaluate::characteristics::Procedure &p,
+                   Fortran::lower::AbstractConverter &c, bool forceImplicit)
+      : CallInterface{c}, proc{p} {
+    if (forceImplicit || proc.CanBeCalledViaImplicitInterface())
+      buildImplicitInterface(proc);
+    else
+      buildExplicitInterface(proc);
+  }
+  bool hasAlternateReturns() const {
+    for (const auto &dummy : proc.dummyArguments)
+      if (std::holds_alternative<
+              Fortran::evaluate::characteristics::AlternateReturn>(dummy.u))
+        return true;
+    return false;
+  };
+  std::string getMangledName() const { return ""; }
+  mlir::Location getCalleeLocation() const { return converter.genLocation(); }
+  Fortran::evaluate::characteristics::Procedure characterize() const {
+    return proc;
+  }
+  bool isMainProgram() const { return false; }
+  const Fortran::evaluate::characteristics::Procedure &
+  getCallDescription() const {
```

Comment:
Without any line breaks on documentation, these methods just sort of blur together.

We should be writing doxygen comments for public interfaces and using some line breaks.

Reply:
Added space and comments